### PR TITLE
[ DO NOT MERGE ] Fix/issue 303 - Global exclude API can take array

### PR DESCRIPTION
**Pull request by RidhamShah**
- `Global exclude` APIs now take an array of users/groups as input
- Changed global exclude `group Id` (groupId + groupType) from database

Other changes

- Solved ContextMetadata throwing error on hard refresh.
- "Enter group type" title is given in UI.
- Change APIs from PUT to POST.
- Make necessary changes in the swagger.
- Change Testcases accordingly (backend & frontend).

### CODE REVIEW

File: frontend/projects/upgrade/src/app/features/dashboard/experiment-users/components/experiment-users/experiment-users.component.ts [hunk 1]
```diff
@@ -99,10 +99,14 @@ export class ExperimentUsersComponent implements OnInit, OnDestroy {
...
-        this.experimentUserService.excludeUser(id);
+        this.experimentUserService.excludeUser([id]);
         break;
       case EntityTypes.GROUP_ID:
-        this.experimentUserService.excludeGroup(id, groupType);
+        const group = {
+          groupId: id,
+          type: groupType
+        };
+        this.experimentUserService.excludeGroup( [group] );
```
Comment (YashilDepani) [inline frontend/projects/upgrade/src/app/features/dashboard/experiment-users/components/experiment-users/experiment-users.component.ts:109]: correct spacing mistake in this line

**Comment by danoswaltCL** [comment]
Is this a breaking API change?

**Comment by RidhamShah** [comment]
> Is this a breaking API change?

NO, mainly globally exclude APIs name are changed and making them send an array of userIds/GroupIds instead of a single userId/GroupId.

File: backend/packages/Upgrade/src/api/services/ExperimentAssignmentService.ts [hunk 2]
```diff
@@ -381,12 +380,14 @@ export class ExperimentAssignmentService {
...
-        return `${type}_${experimentUser.workingGroup[type]}`;
+        return {type: type, groupId: experimentUser.workingGroup[type]};
       });
 
+      const userGroupsToString = this.arrayToString(userGroup);
+
       const [userExcluded, groupExcluded] = await Promise.all([
         this.explicitIndividualExclusionRepository.find({ userId: experimentUser.id }),
-        userGroup.length > 0 ? this.explicitGroupExclusionRepository.find({ where: { id: In(userGroup) } }) : [],
+        userGroup.length > 0 ? await this.explicitGroupExclusionRepository.getExcludedGroups(userGroupsToString) : [],
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/services/ExperimentAssignmentService.ts:390]: Don't use await inside Promise.all

File: backend/packages/Upgrade/src/api/services/ExperimentAssignmentService.ts [hunk 3]
```diff
@@ -1213,4 +1214,13 @@ export class ExperimentAssignmentService {
     }
     return 0;
   }
+
+  private arrayToString(groups: any[]): string {
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/api/services/ExperimentAssignmentService.ts:1218]: Instead of `any` use proper type here

**Review by VivekFitkariwala — CHANGES_REQUESTED** [review]
Added comments for some trivial changes

File: frontend/projects/upgrade/src/app/features/dashboard/experiment-users/components/experiment-users/experiment-users.component.ts (hunk 1 above)
Comment (RidhamShah) [reply frontend/projects/upgrade/src/app/features/dashboard/experiment-users/components/experiment-users/experiment-users.component.ts]: code updated.

File: backend/packages/Upgrade/src/api/services/ExperimentAssignmentService.ts (hunk 2 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/services/ExperimentAssignmentService.ts]: code updated.

File: backend/packages/Upgrade/src/api/services/ExperimentAssignmentService.ts (hunk 3 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/api/services/ExperimentAssignmentService.ts]: code updated.

**Review by VivekFitkariwala — DISMISSED** [review]

File: backend/packages/Upgrade/src/api/controllers/ExcludeController.ts [hunk 4]
```diff
@@ -110,34 +108,40 @@ export class ExcludeController {
...
+   *            description: Authorization Required Error
+   *          '500':
+   *            description: Internal Server Error, Insert Error in database, JSON format is not valid
    */
-  @Put('/user')
-  public excludeUser(@BodyParam('id') id: string): Promise<ExplicitIndividualExclusion> {
-    return this.exclude.excludeUser(id);
+  @Post('/user')
+  public excludeUser(@BodyParam('userIds') userIds: Array<string>): Promise<ExplicitIndividualExclusion[]> {
+    return this.exclude.excludeUser(userIds);
```
Comment (ppratikcr7) [inline backend/packages/Upgrade/src/api/controllers/ExcludeController.ts:134]: @VivekFitkariwala Can we change the function name as well? Like `excludeUsers`. As its not about a single user from now on.

File: backend/packages/Upgrade/src/api/controllers/ExcludeController.ts [hunk 5]
```diff
@@ -194,33 +204,41 @@ export class ExcludeController {
...
+   *          '401':
+   *            description: Authorization Required Error
+   *          '500':
+   *            description: Internal Server Error, Insert Error in database, JSON format is not valid
    */
-  @Put('/group')
-  public excludeGroup(@BodyParam('type') type: string, @BodyParam('id') id: string): Promise<ExplicitGroupExclusion> {
-    return this.exclude.excludeGroup(id, type);
+  @Post('/group')
+  public excludeGroup(@BodyParam('groups') groups: Array<{ groupId: string, type: string }>): Promise<ExplicitGroupExclusion[]> {
```
Comment (ppratikcr7) [inline backend/packages/Upgrade/src/api/controllers/ExcludeController.ts:233]: Same here, we can use `excludeGroups`.

File: backend/packages/Upgrade/src/api/services/ExcludeService.ts [hunk 6]
```diff
@@ -18,8 +18,17 @@ export class ExcludeService {
     return this.explicitIndividualExclusionRepository.find();
   }
 
-  public excludeUser(userId: string): Promise<ExplicitIndividualExclusion> {
-    return this.explicitIndividualExclusionRepository.saveRawJson({ userId });
+  public excludeUser(userIds: Array<string>): Promise<ExplicitIndividualExclusion[]> {
```
Comment (ppratikcr7) [inline backend/packages/Upgrade/src/api/services/ExcludeService.ts:21]: Same. `excludeUsers`?

**Review by ppratikcr7 — COMMENTED** [review]
I have added comments related to changing names of functions that take array of users or groups.

**Comment by amurphy-cl** [comment]
Does this need review?

File: backend/packages/Upgrade/src/api/controllers/ExcludeController.ts (hunk 5 above)
Comment (VivekFitkariwala) [reply backend/packages/Upgrade/src/api/controllers/ExcludeController.ts:233]: The overall url which will be cerated after your change will be
`/explicitExclude/global/excludeGroups`
We already have `explicitExclude` in the path so `/group` seems fine to me.

File: backend/packages/Upgrade/src/api/controllers/ExcludeController.ts (hunk 4 above)
Comment (VivekFitkariwala) [reply backend/packages/Upgrade/src/api/controllers/ExcludeController.ts:134]: +1

**Comment by ppratikcr7** [comment]
> Does this need review?

Yes, we have added some code review.

File: backend/packages/Upgrade/src/api/controllers/ExcludeController.ts (hunk 5 above)
Comment (ppratikcr7) [reply backend/packages/Upgrade/src/api/controllers/ExcludeController.ts:233]: Yes, I am not saying that we change `/group` to `/groups` but we can change the function name from `excludeGroup` to `excludeGroups`

**Review by VivekFitkariwala — APPROVED** [review]

**Pull request closed without merging**